jujugui/static/gui/src/app/state/state.js

```diff
@@ -0,0 +1,283 @@
+      state.store = parts.join('/');
+    }
+    return state;
+  }
+
+  /**
+    Inspects the url path to see if it is for the root.
+    @param {Array} urlParts - The url path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the root value, if any.
+  */
+  _parseRoot(urlParts, state) {
+    State.ROOT_RESERVED.some(key => {
+      if (urlParts[0] === key) {
+        state.root = key;
+        return true;
+      }
+    });
+    return state;
+  }
+
+  /**
+    Parses the search portion of the url without the
+    State.PATH_DELIMETERS.get('search') key value.
+    @param {Array} urlParts - The url path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the search value, if any.
+  */
+  _parseSearch(urlParts, state) {
+    if (urlParts.length > 0) {
+      state.search = urlParts.join('/');
+    }
+    return state;
+  }
+
+  /**
+    Parses the GUI portion of the url without the
+    State.PATH_DELIMETERS.get('gui') key value.
+    @param {Array} urlParts - The url path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object} The updated state to contain the search value, if any.
+  */
+  _parseGUI(urlParts, state) {
+    let indexes = [];
+    State.GUI_PATH_DELIMETERS.forEach(section => {
+      const index = urlParts.indexOf(section);
+      if (index > -1) {
+        indexes.push(index);
+      }
+    });
+    // If there were no sections found then just return.
+    if (!indexes.length) {
+      return state;
+    }
+    // JavaScript array sort sorts alphabetically which causes issues if
+    // you have multiple default sections ie) [0, 9, 17] gets sorted as
+    // [0, 17, 9]. So we must pass a custom sorter to it so that it sorts
+    // numerically.
+    indexes.sort((a, b) => a-b);
+    // Split out and store the sections
+    let guiParts = {};
+    indexes.forEach((index, arIndex) => {
+      const end = indexes[arIndex+1] || urlParts.length;
+      guiParts[urlParts[index]] = urlParts.slice(index+1, end).join('/');
+    });
+    state.gui = guiParts;
+    return state;
+  }
+
+  /**
+    Parses the url and extracts the user delimeted sections.
+    @param {Array} urlParts - The url path split into parts.
+    @param {Object} state - The application state object as being parsed
+      from the URL.
+    @return {Object}
+  */
+  _parseUser(urlParts, state) {
+    // Grab the indexes of the user delimeters as there can be multiple.
+    let indexes = [];
+    urlParts.forEach((item, index) => {
+      if (item === State.PATH_DELIMETERS.get('user')) {
+        indexes.push(index);
+      }
+    });
+    // If we have no user parts then we don't need to parse them.
+    if (indexes.length > 0) {
```

Comment:
Please check if the length === 0 and return earlier, so that we avoid one indentation level easily.